modules/blockbuilder/tenant_store.go

```diff
@@ -104,13 +106,55 @@ func (s *tenantStore) resetHeadBlock() error {
 	return nil
 }
 
-func (s *tenantStore) AppendTrace(traceID []byte, tr *tempopb.Trace, start, end uint32) error {
+func (s *tenantStore) AppendTrace(traceID []byte, tr *tempopb.Trace, startTime time.Time) error {
 	// TODO - Do this async, it slows down consumption
 	if err := s.cutHeadBlock(false); err != nil {
 		return err
 	}
+	start, end := getTraceTimeRange(tr)
+	start, end = s.adjustTimeRangeForSlack(startTime, start, end)
 
-	return s.headBlock.AppendTrace(traceID, tr, start, end)
+	return s.headBlock.AppendTrace(traceID, tr, uint32(start), uint32(end))
+}
+
+func getTraceTimeRange(tr *tempopb.Trace) (startSeconds uint64, endSeconds uint64) {
+	for _, b := range tr.ResourceSpans {
+		for _, ss := range b.ScopeSpans {
+			for _, s := range ss.Spans {
+				if startSeconds == 0 || s.StartTimeUnixNano < startSeconds {
+					startSeconds = s.StartTimeUnixNano
+				}
+				if s.EndTimeUnixNano > endSeconds {
+					endSeconds = s.EndTimeUnixNano
+				}
+			}
+		}
+	}
+	startSeconds = startSeconds / uint64(time.Second)
+	endSeconds = endSeconds / uint64(time.Second)
+
+	return
+}
+
+func (s *tenantStore) adjustTimeRangeForSlack(startTime time.Time, start, end uint64) (uint64, uint64) {
```

Comment:
Since we're adjusting start/end here, we could adjust it to the start and end of the cycle — `[cycleStart-slack , cycleEnd+slack]`. Cycles could be made much longer than the current 5min, and slack would have to increase accordingly.

Reply:
Makes sense to me